go/dbagent/dbagent.go

```diff
@@ -0,0 +1,360 @@
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return Engines, err
+}
+
+func GetTokuDBSize(dbname string) (TokuSize int64, err error) {
+	query := `SELECT SUM(bt_size_allocated) AS tables_size FROM information_schema.TokuDB_fractal_tree_info WHERE table_schema = ?;`
+	err = QueryData(query, sqlutils.Args(dbname), func(m sqlutils.RowMap) error {
+		TokuSize = m.GetInt64("tables_size")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return TokuSize, err
+}
+
+func GetInnoDBLogSize() (InnoDBLogSize int64, err error) {
+	query := `SELECT @@innodb_log_file_size*@@innodb_log_files_in_group AS logFileSize;`
+	err = QueryData(query, sqlutils.Args(), func(m sqlutils.RowMap) error {
+		InnoDBLogSize = m.GetInt64("logFileSize")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return InnoDBLogSize, err
+}
+
+func UserExists(username string) (UserExists bool) {
+	query := `SELECT user FROM mysql.user WHERE user=?;`
+	err := QueryData(query, sqlutils.Args(username), func(m sqlutils.RowMap) error {
+		UserExists = (len(m.GetString("user")) != 0)
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return UserExists
+}
+
+func HasGrant(username string, grant string) (HasGrant bool) {
+	query := `SELECT * FROM mysql.user WHERE user=?;`
+	err := QueryData(query, sqlutils.Args(username), func(m sqlutils.RowMap) error {
+		HasGrant = (m.GetString(grant) == "Y")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return HasGrant
+}
+
+func CreateUser(username string, host string, password string) error {
+	_, err := ExecuteQuery(`CREATE USER IF NOT EXISTS ?@? IDENTIFIED BY ?;`, username, host, password)
+	return err
+}
+
+func GrantUser(username string, host string, grant string, object string, password string) error {
+	var query string
+	if len(password) > 0 {
+		query = fmt.Sprintf("GRANT %s ON %s TO '%s'@'%s' IDENTIFIED BY '%s';", grant, object, username, host, password)
+	} else {
+		query = fmt.Sprintf("GRANT %s ON %s TO '%s'@'%s';", grant, object, username, host)
+	}
+	_, err := ExecuteQuery(query)
+	return err
+}
+
+func ManageReplicationUser() error {
+	var err error
+	version, err := GetMySQLVersion()
+	if err != nil {
+		log.Errore(err)
+	}
+	if !UserExists(config.Config.MySQLReplicationUser) {
+		if version == "5.7" {
+			err = CreateUser(config.Config.MySQLReplicationUser, "%", config.Config.MySQLReplicationPassword)
+			if err != nil {
+				return err
+			}
+		} else {
+			err = GrantUser(config.Config.MySQLReplicationUser, "%", "REPLICATION SLAVE", "*.*", config.Config.MySQLReplicationPassword)
+		}
+	}
+	if !HasGrant(config.Config.MySQLReplicationUser, "Repl_slave_priv") {
+		err = GrantUser(config.Config.MySQLReplicationUser, "%", "REPLICATION SLAVE", "*.*", "")
+	}
+	query := fmt.Sprintf(`FLUSH PRIVILEGES;`)
+	_, err = ExecuteQuery(query)
+	if err != nil {
+		return log.Errore(err)
+	}
+	return err
+}
+
+func GenerateBackupForUsers(users []string) (backup string, err error) {
```

Comment:
Can you please explain what this function does?

Reply:
Sometimes there are tasks when we want to provision for example a read-only replica for some reporting purposes. And we want our reporting team to connect only for this replica, so when provisioning server and mysql we will create a specific login for this purposes. If we then will start provisioning backup from master using xtrabackup, we will lose this login. So this option is added to prevent such cases. There is a special config parameter for such cases called  `"MySQLBackupUsersOnTargetHost": ["'root'@'localhost'","'user_1'@'localhost'","'slave_user_1'@'localhost'","'not_exs_user'@'localhost'"]`